File: all-code/401-500/498findDiagonalOrder.py
# 给你一个大小为 m x n 的矩阵 mat ，请以对角线遍历的顺序，用一个数组返回这个矩阵中的所有元素。
#
#  
#
# 示例 1：
#
#
# 输入：mat = [[1,2,3],[4,5,6],[7,8,9]]
# 输出：[1,2,4,7,5,3,6,8,9]
# 示例 2：
#
# 输入：mat = [[1,2],[3,4]]
# 输出：[1,2,3,4]
#  
#
# 提示：
#
# m == mat.length
# n == mat[i].length
# 1 <= m, n <= 104
# 1 <= m * n <= 104
# -105 <= mat[i][j] <= 105





from typing import List
import random
import bisect
class Solution:

    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        ans = []
        row, col = len(mat), len(mat[0])
        cr, cc = 0, 0
        dir = 1
        for i in range(row * col):
            ans.append(mat[cr][cc])
            if 1 == dir:
                if cc + 1 >= col:
                    cr += 1
                    dir = -dir
                elif cr - 1 < 0:
                    # if cc + 1 < col:
                    #     cc += 1
                    # else:
                    #     cr += 1
                    cc += 1
                    dir = -dir
                else:
                    cr -= 1
                    cc += 1
            else:
                if cr + 1 >= row:
                    cc += 1
                    dir = -dir
                elif cc - 1 < 0:
                    cr += 1
                    dir = -dir
                else:
                    cr += 1
                    cc -= 1
        return ans


so = Solution()

File: c++/src/1-999/498findDiagonalOrder.cpp
#include "lc_pub.h"

class Solution {
public:
    vector<int> findDiagonalOrder(vector<vector<int>>& mat) {
        int r= mat.size(), c=mat[0].size();
        vector<int> ans;
        int dir=0;
        for (int i=0;i<r+c;i++) {
            int x0,y0,xn,yn;
            if (dir==0) {
                x0=min(i, r-1);
                y0=i-x0;
            }
            else {
                y0=min(i, c-1);
                x0=i-y0;
            }
            for (int j=0;j<=i;j++) {
                int x,y;
                if (dir==0) x=x0-j,y=y0+j;
                else x=x0+j,y=y0-j;
                if (0<=x&&x<r&&0<=y&&y<c) 
                    ans.push_back(mat[x][y]);
                else
                    break;
            }
            dir=1-dir;
        }
        return ans;
    }
};

int main()
{
    std::cout<<"test let us start! %s" << __cplusplus <<std::endl;
    auto arr=parseGrid("[[1,2,3],[4,5,6],[7,8,9]]");
    Solution so;
    auto v = so.findDiagonalOrder(arr);
    cout << v << endl;
    return 0;
}
